### ISSUE TITLE
궁금한 점 질문 드립니다..

### ISSUE THREAD

**Issue by mingzzang** [hsnam95/class2019Spring#11]
안녕하십니까, 교수님. 
 
이번학기 교수님의 영어음성학응용 을 수강한 서민지 라고 합니다.
 
다름이 아니오라, 이번 중간고사 이후 내용들을 review하면서 궁금한 부분이 생겨 이렇게 질문 드리게 되었습니다.
 
쓰다 보니 의도치 않게 질문이 많아졌습니다.. 죄송합니다..
 

Q1. boston_housing을 비롯한 인공지능에서 sampling을 할 때,
(x_train, y_train), (x_test, y_test) = tensorflow.keras.datasets.boston_housing.load_data(path='boston_housing.npz')
model.evaluate(x_test, y_test)
 
evaluate 과정에서 안 쓴 20%의 데이터를 쓴다고 했는데, 그렇다면 첫 번째줄의 명령어에서 임의로 3/4과 1/4를 분리 소환시킨 것인가요?
x_train은 404개였고 x_test는 102개 였는데 그럼 원래 데이터는 404+102개 였던 건가요?
어떤 기준으로 20%의 data를 걸러낸 것이고, 어떠한 명령어가 그것을 가능케 한 것인지 갑자기 의문이 들어 질문 드립니다.ㅠㅠ
 

Q2. model.fit => model.evaluate => model.predict 순으로 기억하고 있습니다.
그런데 validation split이 두번째 단계에 해당하는지 세번째 단계에 해당하는지 헷갈립니다.
이름은 validation인데 원래 validation은 마지막에 내가 정한 임의의 수를 넣어보는 단계가 아닌가요? 아니면 test 역할을 하는 두번째 단계인 model.evaluate 단계로 이해해야 하는 것인가요?
 

Q3. mnist 등에서 (60000, ) 식으로 나오는 건 튜플로 굳이 표현해주기 위해서 라고 했는데, 왜 꼭 튜플로 표현해주어야 했나요..?
그리고 imdb에서 (25000, )으로 나온 이유는 영화 리뷰마다 단어 수가 달라서 인가요...?
 
 
Q4. one hot encoding에 대해 print(model.output_shape)에서 (none, 10)에서 10은 무슨 뜻인지 헷갈립니다..ㅠㅠ output이 10차원이라는 의미인가요?
이걸 통해 one hot encoding이 쓰였다고 추론하는 것은 인공지능에서 상식적인 부분인 건가요..?
추가로 이걸 그림으로 생각해봤을 때, 마지막에 동그라미가 10개면, 그 각 10개 중 하나에만 화살표가 감으로써 표현되는 건가요, 아니면 각 원에 1000.. 0100.. 이 의미가 있는 건가요..?
 

Q5. 마지막으로.. 4번에 이어서 헷갈리기 시작한 문제 같습니다... mnist와 같은 파일의 output은 (none, 10) 으로 표현이 되는데,  왜 boston_housing이나 imdb의 경우에는 1*1을 맞추어 (none, 1)의 형태로 만들어 주어야 하나요..?
결국 모두 입력 값을 넣었을 때 하나의 출력 값(집값/숫자와 같은 예측값/좋다or나쁘다)만을 나타내야 하는 건 똑같지 않은가요...? 굳이 mnist만 2진법을 사용하여 10으로 나타낸 이유가 무엇인가요..? ㅠㅠ

**Comment by cjw414** [comment]
1에 대한 답변입니다.

우선 x_train은 404개, x_test는 102개로 질문하신 대로 원래 데이터는 404+102개가 맞습니다.

그리고 test_split(20%)에 대해서는  텐서플로우 공식 문서에서의 tf.keras.datasets.boston_housing.load_data에 설명되어 있는데, 다음과 같이 설명되어 있습니다.
![image](https://user-images.githubusercontent.com/44089395/59328043-5bcdd300-8d26-11e9-996a-dde7b6fd6843.png)

여기서 test의 비율이 되는 argument는 test_split인데, 이에 대한 값을 넣어주지 않으면 default 값으로 0.2가 들어가는 것을 확인할 수 있습니다.

**Comment by cjw414** [comment]
Q2에 대한 답변입니다.
model.fit(), model.evaluate(), model.predict()에 대한 개념부터 짚고 넘어가면 좋을 거 같습니다.
우선 model.fit()은 말 그대로 저희가 만든(model.compile()) 모델을 준비한 train data에 맞게 훈련을 시키는 것입니다.
model.evaluate()는 훈련을 끝낸 모델을 저희가 준비한 test_data에 대해서 다양한 지표(loss,accuracy 등이 있지만 이 부분에 대해서는 우선은 너무 깊게 안 가도 될 거 같습니다) '평가(evaluate)'하는 것입니다.

마지막으로 model.predict()는 훈련된 모델에 test dataset이 아닌, 임의의 input sample을 주었을 때 출력값을 내뱉도록 하는 것입니다.

이에 대한 더욱 자세한 설명은 텐서플로우나 케라스 공식 문서를 참조하시면 될 것 같고(https://keras.io/models/model/#methods)

validation dataset은 훈련 도중 매 epoch가 끝날 때마다 모델의 성능을 알아보고자 따로 준비한, train/test와는 별개의 제 3의 dataset입니다.
validation은 훈련 도중에 일어나기 때문에 model.fit()에 들어가는 것이 자연스럽다고 할 수 있습니다.
다음은 validation_split이 사용된 예입니다.
history = model.fit(x_train, y_train, epochs = 3, validation_split=0.2)

**Comment by cjw414** [comment]
Q3에 대한 답변입니다.
해당 값들은 y_train.shape, y_test.shape와 같은 값입니다.
그리고 이렇게 빈 tuple로 나타내는 이유는 np array(혹은 tf.tensor)의 shape가 다음과 같이 정의돼 있기 때문입니다.(https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.shape.html)
ndarray.shape : Tuple of array dimensions.

따라서 60000개의 숫자가 있더라도 (60000, )이런 식으로, tuple이 나와야 합니다.

**Comment by cjw414** [comment]
Q4에 대한 답변입니다.(Q5부터 보고 오시는 걸 추천드립니다.)
질문하신 대로 (None, 10)에서의 10은 output의 차원을 의미합니다. 10은 저희가 추측하고자 하는 대상의 class의 개수라고 생각하시면 됩니다.

mnist의 경우 0~9가 각각 [1 0 0 0 0 0 0 0 0 0 0] ~ [0 0 0 0 0 0 0 0 0 0 1]로 저희의 class만큼의 차원에서의 one hot vector로 표현되는 것입니다.

동그라미로 이해를 하고자 하시면, 다음과 같은 그림으로 이해하시면 됩니다.

![image](https://user-images.githubusercontent.com/44089395/59330374-3fcd3000-8d2c-11e9-8f6e-489ee06823d1.png)

여기서 직관적으로 이해하기 쉽게 설명하자면 마지막에 나오는 숫자 값은 각각 우리가 input으로 넣은 이미지가 0일 확률, 1일 확률, 2일 확률... 9일 확률임을 나타내는 것이다.

**Comment by cjw414** [comment]
Q5에 대한 답변입니다.
이처럼 node가 다르게 되는 것은 각각 모델의 task가 다르기 때문입니다.
boston housing의 경우 '집값'이라는 연속된 값(ex.$100~$300)을 예측하는 regression model을 만들게 됩니다.

이에 반해 imdb나 mnist는 classification task입니다. imdb에서 해당 comment의 '긍정'/ '부정'을 예측하는 task는 불연속적인 값(ex. 남/여)을 분류하는 task로 이를 classification이라고 합니다. imdb의 경우 긍/부정만을 분류하므로 binary classification이라 하며 이 때문에 마지막에 1개의 node(동그라미)만 있는 것입니다.
나아가 mnist는 티셔츠/바지/스웨터/신발...처럼 다양한 class에 대해 분류하는 작업인데, 이와 같은 경우 저희가 분류하고 싶은 class 만큼의 노드(mnist에서는 10개)가 있는 것입니다.

**Comment by mingzzang** [comment]
덕분에 5개 다 이해했습니다 감사드립니다ㅠㅠㅠㅠ
